.github/scripts/toolchain/group-gcc-test-failures.py

```diff
@@ -0,0 +1,149 @@
+        return filepath.split("/")
+    
+    @dataclasses.dataclass
+    class PathSegmentStats:
+        count: int = 0
+        children: dict[str, "PathSegmentStats"] = dataclasses.field(default_factory=dict)
+
+    def print_path_segment_stats(
+        entry: PathSegmentStats,
+        print_limit: int = 0,
+        print_markdown_list: bool = False,
+        indent: int = 0
+    ) -> None:
+        """Prints hierarchical path segments with their occurrence counts
+        
+        print_limit: Print only path segments with occurrence count >= print_limit
+        """
+        sorted_path_stats = sorted(entry.children.items(), key=lambda x: x[1].count, reverse=True)
+        for key, value in sorted_path_stats:
+            if value.count < print_limit:
+                if not print_markdown_list:
+                    print(" " * (indent) + "...")
+                return
+            
+            if print_markdown_list:
+                print(" " * indent + f"- {key}: {value.count}")
+            else:
+                print(" " * indent + f"{key}: {value.count}")
+            print_path_segment_stats(value, print_limit, print_markdown_list, indent + 2)
+
+    def update_top_counts(entry: PathSegmentStats, sorted_top_counts: list[int]) -> None:
+        """Updates recursively the sorted_top_counts list with the top counts from the entry"""
+        for segment_stats in entry.children.values():
+            if segment_stats.count > sorted_top_counts[0]:
+                sorted_top_counts[0] = segment_stats.count
+                sorted_top_counts.sort()
+                update_top_counts(segment_stats, sorted_top_counts)
+
+    root_seg_stats = PathSegmentStats()
+    for messages in log_messages.values():
+        for message in messages:
+            path_segments = parse_filepath(message)
+            curr_seg_stats = root_seg_stats
+            curr_seg_stats.count += 1
+            for segment in path_segments:
+                if segment not in curr_seg_stats.children:
+                    curr_seg_stats.children[segment] = PathSegmentStats()
+                curr_seg_stats = curr_seg_stats.children[segment]
+                curr_seg_stats.count += 1
+
+    TOP_SEGMENTS = 10
+    top_counts = [0] * TOP_SEGMENTS
+    update_top_counts(root_seg_stats, top_counts)
+    with open(summary_output, "a") as summary_output_file:
+        with contextlib.redirect_stdout(summary_output_file):
+            print(f"## Top {TOP_SEGMENTS} path segments:")
+            print_path_segment_stats(root_seg_stats, print_limit=min(top_counts), print_markdown_list=True)
+
+    print_path_segment_stats(root_seg_stats, print_limit=100)
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(description="Aggregate test results from log files")
+    parser.add_argument("--dir", required=True, type=str, help="Directory containing log files")
+    parser.add_argument("--summary_output", required=True, type=str, help="Path to the summary output file")
```

Comment:
Do I understand correctly that this script prints prints the detailed stats to the stdout while the summary to the specified file? This seems quite inconsistent. I'd rather propose to either write the summary to the stdout as well or having a separate script that would list top issues from the previously obtained output.

Reply:
Yes, it works as you described. Since we want to store detailed output in artefacts and write summary to `GITHUB_STEP_SUMMARY`, it seemed more future-proof to me to split summary and detailed output inside the script. I think the only other option how to split them is to write bash script which would parse output and find summary in it. I agree that current approach makes script output inconsistent but it seemed better to me overall because with parsing bash script you'd need to change code on 2 places if you want to change the output. Do you think otherwise, that the approach with separate summary parsing script is prefereable?

Reply:
If your priority is to keep the code at one place, what about having two output modes for the script: details and summary? The script would be then run twice, once for details and once for the summary.

Reply:
Good idea, since the script execution is fast, this seems like the best option to me.

Reply:
At the end I removed "top10" from `group-gcc-test-failures.py` and rewrote it as you suggested as a different algorithm which only works with strings in `extract-most-frequent-failures.sh`